src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -63,5 +78,39 @@ internal override async Task<IPackageSearchMetadata> GetPackageMetadataAsync<T>(
 
             return await base.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
         }
+
+        internal static PackageCollectionItem GetLatestApplicableTransitivePackageVersion(IGrouping<string, PackageCollectionItem> groupedPackagesById, IEnumerable<PackageCollectionItem> installedPkgs)
+        {
+            if (groupedPackagesById == null || !groupedPackagesById.Any())
+            {
+                throw new ArgumentException(Strings.Argument_Cannot_Be_Null_Or_Empty, nameof(groupedPackagesById));
+            }
+            if (installedPkgs == null)
+            {
+                throw new ArgumentNullException(nameof(installedPkgs));
+            }
+            if (groupedPackagesById
+                    .Any(group => !group.PackageReferences.Any() // PackageReferences is empty
+                        || group.PackageReferences.Any(pkgRef =>
+                            pkgRef is not ITransitivePackageReferenceContextInfo tr // it is not a transitive package
+                            || !tr.TransitiveOrigins.Any()))) // No transitive origin for a transitive package
+            {
+                throw new ArgumentException(Strings.InvalidCollectionElementType, nameof(groupedPackagesById));
+            }
+
+            IEnumerable<PackageCollectionItem> matchedOrigins = groupedPackagesById
+                .Where(group => group.PackageReferences
+                    .Cast<ITransitivePackageReferenceContextInfo>()
+                    .Any(trPkgRef => trPkgRef.TransitiveOrigins.Any(trOrigin => installedPkgs.Contains(trOrigin.Identity))));
```

Comment:
If installed packages are many(`n`) and transitive grouped transitive packages are(`m`) then it would be `n*m` string lookup going to happen.  I'm expecting CPU calculation would go up. Is there any measurement you're tracking for noticeable regression?

Reply:
To avoid similar expensive lookup we implemented [Trie algorithm here](https://github.com/NuGet/NuGet.Client/blob/4c236cd40ee329c1b8b3d26fa16d5f37598ab605/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/SearchTree.cs#L83). Maybe you don't have to do now, but create a tracking issue and implement in the future if this becomes bottlenecked.